src/commands/run.rs

```diff
@@ -0,0 +1,177 @@
+use super::CommandTrait;
+use crate::{
+    cli::{create_path_env, Config},
+    error::{MainError, PnError},
+    workspace,
+};
+use pipe_trait::Pipe;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::ffi::OsString;
+use std::fs::File;
+use std::num::NonZeroI32;
+use std::path::Path;
+use std::{process, process::Stdio};
+
+/// Runs a defined package script.
+#[derive(Debug, clap::Args)]
+#[clap(rename_all = "kebab-case")]
+pub struct Run {
+    /// Name of the package script to run.
+    pub script: String, // Not OsString because it would be compared against package.json#scripts
+
+    /// Arguments to pass to the package script.
+    pub args: Vec<OsString>,
+}
+
+impl CommandTrait for Run {
+    fn run(self, config: Config) -> Result<(), MainError> {
+        let mut cwd = std::env::current_dir().expect("Couldn't find the current working directory");
+        if config.workspace_root {
+            cwd = workspace::find_workspace_root(&cwd)?;
+        }
+        let manifest_path = cwd.join("package.json");
+        let manifest = read_package_manifest(&manifest_path)?;
+        if let Some(command) = manifest.scripts.get(&self.script) {
+            eprintln!(
+                "\n> {pkg_name}@{pkg_version} {script} {path}",
+                pkg_name = &manifest.name,
+                pkg_version = &manifest.version,
+                script = &self.script,
+                path = cwd.display(),
+            );
+            eprintln!("> {command}\n");
+            return run_script(&self.script, command, &cwd);
+        } else {
+            let e: Result<(), MainError> = PnError::MissingScript { name: self.script }
+                .pipe(MainError::Pn)
+                .pipe(Err);
+            return e;
+        }
+    }
+}
+
+/// Structure of `package.json`.
+#[derive(Debug, Clone, Deserialize, Serialize)]
+#[serde(rename_all = "kebab-case")]
+pub struct NodeManifest {
+    #[serde(default)]
+    name: String,
+
+    #[serde(default)]
+    scripts: HashMap<String, String>,
+
+    #[serde(default)]
+    version: String,
```

Comment:
Both `pnpm run` and `npm run` functions without `name` and `version` in `package.json`. There is no reason to make them mandatory field.

Reply:
Is there any reason `name` and `version` are not next to each other?

Reply:
there actually is. initally i did put them side by side but the read_package_manifest test was failing. the json! macro ends up sorting the `expected` value alphabetically while parsing the nodemanifest struct to json sorts it in the same way as the struct is defined. so the test ends up failing. im unusure how else it could be changed. we could maybe only compare the scripts hashmap with another hashmap or something.

Reply:
Heres an image of the test. the first one is the `received` json and the second is `expected`
![image](https://github.com/pnpm/pn/assets/91675932/62c3b316-4fe6-4caf-ba7c-742076341da9)

thus it failed
![image](https://github.com/pnpm/pn/assets/91675932/bf0ba1aa-22e4-4ec8-a6fa-07c17f43b3a4)